### ISSUE TITLE
RoundUp behavior on integer funcs can cause arithmetic exceptions

### ISSUE THREAD

**Issue by alexreinking** [halide/Halide#4423]
Consider the two pipelines

    output_i(x) = 100 / (27 - x)
    output_f(x) = 100.f / (27.f - x)

Then in either case writing

    output_*.split(x, xo, xi, 8, TailStrategy::RoundUp);

If you then try to realize via:

    output_*.realize(24);

In either case everything is fine. But if you try to realize via:

    output_*.realize(25);

Then `output_i` crashes with a floating point exception (ie. integer division by zero) while `output_f` returns a buffer containing the correct values (because FP div-by-zero doesn't trap).

This behavior is slightly unexpected since the compute bounds don't contain the problematic value. While this particular example is designed to be pathological, here is a program that actually runs into this issue:

```
#include <Halide.h>
using namespace Halide;

class RoundUp : public Generator<RoundUp>
{
public:
    Output<Buffer<int>> output{"output", 1};

    void generate() {
        Func f, g;
        Var x, xo, xi;

        g(x) = 5;
        f(x) = 100 / g(x);
        output(x) = f(x);

        g.compute_root();
        f.compute_root();
        f.split(x, xo, xi, 8, TailStrategy::RoundUp);
    }
};

HALIDE_REGISTER_GENERATOR(RoundUp, round_up)
```

I get:

```
[alex@reinking-arch round_up]$ ./run_round_up output=foo.mat --output_extents=[25]
Floating point exception (core dumped)
```

One fix would be to reject this tail strategy for rules that divide by an integer which contains zero in the range.

**Comment by alexreinking** [comment]
Note that this issue also arises with multiplication / division by -1 when the other value is `INT_MIN`

**Comment by abadams** [comment]
Yeah, I've seen this pop up in practice. I think we just need to define
division by zero as returning zero.

**Comment by alexreinking** [comment]
How will that work in the implementation? Won't that make integer division by an unknown less efficient? JRK and I briefly looked online to see if there was any integer divide by zero that didn't trap and we couldn't find one.

**Comment by abadams** [comment]
Yes, that would make integer division by an unknown that can't statically be proven to be not zero less efficient. Fortunately that doesn't seem to happen much in image processing. Division by non-constant integers is very rare to begin with, and when it happens it usually includes an explicit guard against zero, e.g. x / max(y, 1), so we can use a if (can_prove(denominator != 0)) to skip the extra logic.

Note that Halide signed integer division is already more than a native div instruction, due to the rounding difference.

**Comment by abadams** [comment]
Note that we also have the quiet_div intrinsic, which returns an llvm UndefValue when the denominator is zero instead of trapping. That could turn into UB really easily though, so I don't think that's a good general plan.

**Comment by alexreinking** [comment]
@abadams 

> That could turn into UB really easily though, so I don't think that's a good general plan.

I agree. Making integer division total makes for more predictable behavior. As long as arithmetic errors don't trap, RoundUp should be safe

**Comment by alexreinking** [comment]
This was fixed by #4439
